comp/collector/collector/collectorimpl/collector.go

```diff
@@ -75,3 +104,255 @@ func newNoneCollector() provides {
+	}
+
+	pkgCollector.InitPython(common.GetPythonPaths()...)
+
+	deps.Lc.Append(fx.Hook{
+		OnStart: c.start,
+		OnStop:  c.stop,
+	})
+
+	c.log.Debug("Collector up and running!")
+	return c
+}
+
+// AddEventReceiver adds a callback to the collector to be called each time a check is added or removed.
+func (c *collectorImpl) AddEventReceiver(cb collector.EventReceiver) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	c.eventReceivers = append(c.eventReceivers, cb)
+}
+
+func (c *collectorImpl) notify(cid checkid.ID, e collector.EventType) {
+	for _, cb := range c.eventReceivers {
+		cb(cid, e)
+	}
+}
+
+// start begins the collector's operation.  The scheduler will not run any checks until this has been called.
+func (c *collectorImpl) start(_ context.Context) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	run := runner.NewRunner(c.senderManager)
+	sched := scheduler.NewScheduler(run.GetChan())
+
+	// let the runner some visibility into the scheduler
+	run.SetScheduler(sched)
+	sched.Run()
+
+	c.scheduler = sched
+	c.runner = run
+	c.state.Store(started)
+
+	return nil
+}
+
+// stop halts any component involved in running a Check
+func (c *collectorImpl) stop(_ context.Context) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if c.scheduler != nil {
+		c.scheduler.Stop() //nolint:errcheck
+		c.scheduler = nil
+	}
+	if c.runner != nil {
+		c.runner.Stop()
+		c.runner = nil
+	}
+	c.state.Store(stopped)
+	return nil
+}
+
+// RunCheck sends a Check in the execution queue
+func (c *collectorImpl) RunCheck(inner check.Check) (checkid.ID, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	ch := middleware.NewCheckWrapper(inner, c.senderManager)
+
+	var emptyID checkid.ID
+
+	if c.state.Load() != started {
+		return emptyID, fmt.Errorf("the collector is not running")
+	}
+
+	if _, found := c.checks[ch.ID()]; found {
+		return emptyID, fmt.Errorf("a check with ID %s is already running", ch.ID())
+	}
+
+	err := c.scheduler.Enter(ch)
+	if err != nil {
+		return emptyID, fmt.Errorf("unable to schedule the check: %s", err)
+	}
+
+	// Track the total number of checks running in order to have an appropriate number of workers
+	c.checkInstances++
+	if ch.Interval() == 0 {
+		// Adding a temporary runner for long running check in case the
+		// number of runners is lower than the number of long running
+		// checks.
+		c.log.Infof("Adding an extra runner for the '%s' long running check", ch)
+		c.runner.AddWorker()
+	} else {
+		c.runner.UpdateNumWorkers(c.checkInstances)
+	}
+
+	c.checks[ch.ID()] = ch
+	c.notify(ch.ID(), collector.CheckRun)
+	return ch.ID(), nil
+}
+
+// StopCheck halts a check and remove the instance
+func (c *collectorImpl) StopCheck(id checkid.ID) error {
+	if !c.started() {
+		return fmt.Errorf("the collector is not running")
+	}
+
+	ch, found := c.get(id)
+	if !found {
+		return fmt.Errorf("cannot find a check with ID %s", id)
+	}
+
+	// unschedule the instance
+	err := c.scheduler.Cancel(id)
```

Comment:
I'd suggest combining the `err != nil` check with this line, and same for the next two, since you don't need to use the `err` for anything aside from wrapping and returning it.